freqtrade/freqai/base_models/PyTorchModelTrainer.py

```diff
@@ -0,0 +1,184 @@
+        :param batch_size: The size of the batches to use during training.
+        :param max_iters: The number of training iterations to run.
+            iteration here refers to the number of times we call
+            self.optimizer.step(). used to calculate n_epochs.
+        :param eval_iters: The number of iterations used to estimate the loss.
+        :param init_model: A dictionary containing the initial model/optimizer
+         state_dict and model_meta_data saved by self.save() method.
+        :param model_meta_data: Additional metadata about the model (optional).
+        """
+        self.model = model
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.model_meta_data = model_meta_data
+        self.device = device
+        self.max_iters = max_iters
+        self.batch_size = batch_size
+        self.eval_iters = eval_iters
+
+        if init_model:
+            self.load_from_checkpoint(init_model)
+
+    def fit(self, data_dictionary: Dict[str, pd.DataFrame]):
+        """
+         - Calculates the predicted output for the batch using the PyTorch model.
+         - Calculates the loss between the predicted and actual output using a loss function.
+         - Computes the gradients of the loss with respect to the model's parameters using
+           backpropagation.
+         - Updates the model's parameters using an optimizer.
+        """
+        data_loaders_dictionary = self.create_data_loaders_dictionary(data_dictionary)
+        epochs = self.calc_n_epochs(
+            n_obs=len(data_dictionary["train_features"]),
+            batch_size=self.batch_size,
+            n_iters=self.max_iters
+        )
+        for epoch in range(epochs):
+            # evaluation
+            losses = self.estimate_loss(data_loaders_dictionary, data_dictionary)
+            logger.info(
+                f"epoch ({epoch}/{epochs}):"
+                f" train loss {losses['train']:.4f} ; test loss {losses['test']:.4f}"
+            )
+            # training
+            for batch_data in data_loaders_dictionary["train"]:
+                xb, yb = batch_data
+                xb = xb.to(self.device)
+                yb = yb.to(self.device)
+                yb_pred = self.model(xb)
+                loss = self.criterion(yb_pred, yb)
+
+                self.optimizer.zero_grad(set_to_none=True)
+                loss.backward()
+                self.optimizer.step()
+
+    @torch.no_grad()
+    def estimate_loss(
+            self,
+            data_loader_dictionary: Dict[str, DataLoader],
+            data_dictionary: Dict[str, pd.DataFrame]
+    ) -> Dict[str, float]:
+
+        self.model.eval()
+        epochs = self.calc_n_epochs(
+            n_obs=len(data_dictionary["test_features"]),
+            batch_size=self.batch_size,
+            n_iters=self.eval_iters
+        )
+        loss_dictionary = {}
+        for split in ["train", "test"]:
```

Comment:
Unless I am missing something, you appear to compute loss on the entire train dataset and then again, in the subsequent loop (line 76). These values will be the same wont they? So it appears to be wasted cycles. 

Seems to me you only need to do `test` here.

Reply:
inside the training loop the model parameters are being updated every batch iteration.
so, the loss there is'nt representing the loss on the entire data.
i'm updating the eval iteration to go through all the train/test data once.

Reply:
Ok I see. In that case, what is the use of that train loss in comparison to simply reporting the  sum of train loss encountered during the most recent epoch?

I agree that the test loss important to do in this fashion - I am more concerned about the value of this train loss. Let me know how you would use that value and how it would help you more than just using epoch train loss summation and validation loss reporting.

Reply:
just for comparing the train to test loss. no other uses.
the difference between what you suggest to what i did will only be relevant in big datasets.
we can use the last epoch.